### PR TITLE
Use the engine interface with AFL on a trial basis.

### PR DESCRIPTION
**Pull request by mbarbella-chromium**
Not yet ready for review. This still needs a bit of cleanup and much more careful testing.

### CODE REVIEW

**Comment by mbarbella-chromium** [comment]
Still needs a small amount of work to get tests passing. Feel free to take a look now if you want, but I'll update this again once everything is green.

**Comment by mbarbella-chromium** [comment]
Tests are now passing.

File: src/python/bot/fuzzers/afl/engine.py [hunk 1]
```diff
@@ -12,13 +12,35 @@
...
 from bot.fuzzers.afl import launcher
+from bot.fuzzers.afl import stats
 from lib.clusterfuzz.fuzz import engine
+from system import environment
+
+
+def _get_command_from_fuzz_result(fuzz_result, runner):
+  """Get the command string without showing unnecessary minijail commands."""
+  command = fuzz_result.command
+  if environment.get_value('USE_MINIJAIL'):
```
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:28]: nit: MINIJAIL is now completely disabled, so can skip this.

File: src/python/bot/fuzzers/afl/engine.py [hunk 2]
```diff
@@ -82,13 +135,21 @@ def reproduce(self, target_path, input_path, arguments, max_time):
     Returns:
       A ReproduceResult.
     """
-    config = launcher.AFLConfig.from_target_path(target_path)
+    environment.set_value('AFL_ENGINE_TIMEOUT', max_time)
```
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:138]: this doesn't do anything in the reproduce case right?

File: src/python/bot/fuzzers/afl/engine.py [hunk 3]
```diff
@@ -50,25 +73,55 @@ def fuzz(self, target_path, options, reproducers_dir, max_time):
...
-          are found.
+      reproducers_dir: The directory to put reproducers in when crashes are
+        found.
       max_time: Maximum allowed time for the fuzzing to run.
 
    Returns:
       A FuzzResult object.
     """
-    config = launcher.AFLConfig.from_target_path(target_path)
+    environment.set_value('AFL_ENGINE_TIMEOUT', max_time)
```
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:83]: rather than set an env var here to be used later, can we just modify runner's fuzz method to accept this argument, which is then passed to afl's get_fuzz_timeout?

File: src/python/bot/fuzzers/afl/engine.py [hunk 4]
```diff
@@ -50,25 +73,55 @@ def fuzz(self, target_path, options, reproducers_dir, max_time):
...
-    raise NotImplementedError
+    fuzz_result = runner.fuzz()
+
+    command = _get_command_from_fuzz_result(fuzz_result, runner)
+    time_executed = fuzz_result.time_executed
+    logs = fuzz_result.output.splitlines() + runner.fuzzer_stderr.splitlines()
+
+    # Bail out if AFL returns a nonzero status code.
+    if fuzz_result.return_code:
+      return engine.FuzzResult(logs, command, [], {}, time_executed)
```
Comment (oliverchang) [inline src/python/bot/fuzzers/afl/engine.py:104]: nit: log_error to match libFuzzer: https://github.com/google/clusterfuzz/blob/0b5c023eca9e3aac41faba17da8f341c0ca2ddc7/src/python/bot/fuzzers/libFuzzer/engine.py#L283

File: src/python/bot/tasks/fuzz_task.py [hunk 5]
```diff
@@ -1900,12 +1900,19 @@ def run(self):
...
+
     # TODO(mbarbella): Remove the environment variable check when the old
     # blackbox pipeline has been converted to the engine interface.
-    engine_impl = engine.get(self.fuzzer.name)
     if not engine_impl and environment.get_value('USE_BLACKBOX_ENGINE'):
       engine_impl = engine.get('blackbox')
 
+    # TODO(mbarbella): Remove this check once AFL is fully converted to the new
+    # engine interface.
+    if (self.fuzzer.name == 'afl' and
```
Comment (oliverchang) [inline src/python/bot/tasks/fuzz_task.py:1912]: I'm guessing the plan is to turn this on for reproduction too once we verify that fuzzing seems to work?

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    options = engine_impl.prepare(CORPUS_DIRECTORY, fuzzer_path, DATA_DIRECTORY)
+
+    result = engine_impl.fuzz(fuzzer_path, options, OUTPUT_DIRECTORY,
+                              FUZZ_TIMEOUT)
+
+    self.assertEqual('{0}/afl-fuzz'.format(DATA_DIRECTORY), result.command[0])
+    self.assertIn('-i{0}'.format(CORPUS_DIRECTORY), result.command)
+
+    # Ensure that we've added something other than the dummy file to the corpus.
+    self.assertTrue(os.listdir(CORPUS_DIRECTORY))
```
Comment (oliverchang) [inline src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:83]: Can we move/copy some more tests from afl_launcher_integration_test here which tests that we are actually catching crashes properly?

**Review by oliverchang — COMMENTED** [review]
nice! some initial comments.

File: src/python/bot/fuzzers/afl/engine.py (hunk 1 above)
Comment (jonathanmetzman) [reply src/python/bot/fuzzers/afl/engine.py:28]: Really? What replaced it for misbehaved OSS-Fuzz targets like rar?

File: src/python/bot/fuzzers/afl/engine.py (hunk 3 above)
Comment (jonathanmetzman) [reply src/python/bot/fuzzers/afl/engine.py:83]: +1

File: src/python/bot/fuzzers/afl/engine.py (hunk 2 above)
Comment (jonathanmetzman) [reply src/python/bot/fuzzers/afl/engine.py:138]: +1

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    afl_launcher_integration_test.dont_use_strategies(self)
+
+  def tearDown(self):
+    clear_temp_dir()
+
+  def test_fuzz(self):
+    """Test for fuzz."""
+    engine_impl = engine.AFLEngine()
+
+    _ = afl_launcher_integration_test.setup_testcase_and_corpus(
```
Comment (jonathanmetzman) [inline src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:71]: nit: Don't even assign if we aren't using the result.

File: src/python/bot/fuzzers/afl/engine.py (hunk 1 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/afl/engine.py:28]: No replacement. @inferno-chromium disabled it and was monitoring.

File: src/python/bot/fuzzers/afl/engine.py (hunk 4 above)
Comment (mbarbella-chromium) [reply src/python/bot/fuzzers/afl/engine.py]: Done!

File: src/python/bot/fuzzers/afl/engine.py (hunk 2 above)
Comment (mbarbella-chromium) [reply src/python/bot/fuzzers/afl/engine.py]: Right, removed.

File: src/python/bot/tasks/fuzz_task.py (hunk 5 above)
Comment (mbarbella-chromium) [reply src/python/bot/tasks/fuzz_task.py:1912]: We should actually do this now. Made that change as well. It'll be gated behind a separate job type initially, so there's no harm in enabling in both cases.

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py (hunk 7 above)
Comment (mbarbella-chromium) [reply src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py]: Done!

File: src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py (hunk 6 above)
Comment (mbarbella-chromium) [reply src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:83]: Sure, I've added one for testing a crash case. That seems to be failing, but I'm working on a fix for that now. Just publishing these comments in the meantime.

For the others, whenever we stop using the launcher we should make sure all of those tests are preserved. This is essentially just calling into the launcher, so I'm satisfied that so long as we know that we can handle the basic cases (adding stuff to the corpus, processing crashes, reproducing) that the corner cases tested there will function properly.

Comment (mbarbella-chromium) [reply src/python/tests/core/bot/fuzzers/afl/afl_engine_test.py:86]: Added another new test, and both are now passing.

File: src/python/bot/testcase_manager.py [hunk 8]
```diff
@@ -577,6 +577,12 @@ def __init__(self,
     else:
       engine_impl = None
 
+    # TODO(mbarbella): Remove this check once AFL is fully converted to the new
+    # engine interface.
+    if (fuzz_target and fuzz_target.engine == 'afl' and
+        not environment.get_value('USE_AFL_ENGINE')):
+      engine_impl = None
```
Comment (oliverchang) [inline src/python/bot/testcase_manager.py:584]: Can just put this in engine.get() instead of duplicating the checks in two places.

File: src/python/bot/tasks/fuzz_task.py (hunk 5 above)
Comment (oliverchang) [reply src/python/bot/tasks/fuzz_task.py:1920]: Ok, we need to be very careful here to make sure existing testcases don't break. Please verify this manually on a number of testcases after landing (prior to our progression cron).

File: src/python/bot/tasks/fuzz_task.py [hunk 9]
```diff
@@ -1904,12 +1904,23 @@ def run(self):
           'Unable to setup data bundle %s.' % self.fuzzer.data_bundle_name)
       return
 
+    # For some binaries, we specify trials, which are sets of flags that we only
+    # apply some of the time. Adjust APP_ARGS for them if needed.
+    trials.setup_additional_args_for_app()
```
Comment (oliverchang) [inline src/python/bot/tasks/fuzz_task.py:1909]: What is this for?

**Review by oliverchang — APPROVED** [review]
Thanks. Some additional comments, but LGTM overall. Please test this carefully.

File: src/python/bot/tasks/fuzz_task.py (hunk 9 above)
Comment (mbarbella-chromium) [reply src/python/bot/tasks/fuzz_task.py]: This was incorrectly included when resolving a conflict. Removed.

File: src/python/bot/testcase_manager.py (hunk 8 above)
Comment (mbarbella-chromium) [reply src/python/bot/testcase_manager.py]: Good call. Done.

**Comment by mbarbella-chromium** [comment]
I'll hold off on landing this until after the progression cron runs to give myself a bit more time for manual testing. All comments should be addressed, and tests are all passing.

**Comment by mbarbella-chromium** [comment]
Spoke to Oliver via chat. We'll just land this now and monitor.
